design/service-library.md

```diff
@@ -0,0 +1,187 @@
+```
+
+### Calling Microsoft Graph via the Graph JavaScript Service Library
+
+The core value of the Microsoft Graph JavaScript Service Library is the availability of the chaned request builder pattern (fluent API). This provides developers with typechecking and discoverability when building on Microsoft Graph. This would highlight the full spectrum of capabilities on Microsoft Graph directly in the API Client.
+
+Models are also available in this package and should reflect the underlying version of Graph we are targeting. Developers using our types should do it in a way that will be familiar to how they are used to with other API Clients and APIs. We should provide an easy-to-use model that feels natural in the JavaScript world. This means we should not be forcing developers to use Classes and / or other structures that will make their code less natural to them.
+
+Because of the richness of the Graph APIs, we are making a decision to require developers to selectively import the areas of Graph they care about. This will provide a lightweight bundle size while still offering a good developer experience.
+
+### Selective Imports of the Graph areas needed
+
+To offer a bundle size that fair to the developer, we are using a selective imports strategy. This strategy enables developers to bring only the endpoints they need from Graph. Developers would only see the root paths of Graph when using the API Client by default, and would need to import additional path with selective imports.
+
+#### Using the service library without any selective imports
+
+> ❌ **Fails** <br />
+> This example does not work because `usersById().directReports` is not available at the root of the Graph Client
+
+```typescript
+const directReports = await graphClient.usersById("ab677c1b-50c6-4013-aa9f-792ed42c59c8").directReports.get();
+```
+
+> ✅ **Succeeds** <br />
+> This example succeeds because `me` is at the root of the Graph Client
+
+```typescript
+const me = await graphClient.me.get();
+```
+> ✅ **Succeeds** <br />
+> This example succeeds because `users` is at the root of the Graph Client
+
+```typescript
+const users = await graphClient.users.get();
+```
+
+#### Using the service library with selective imports
+
+> ❌ **Fails** <br />
+> This example fails because `@microsoft/microsoft-graph-client/users/item` is not imported and therefore not available on the Graph Client
+
+```typescript
+import "@microsoft/microsoft-graph-client/users";
+const directReports = await graphClient.usersById("ab677c1b-50c6-4013-aa9f-792ed42c59c8").directReports.get();
+```
+
+> ✅ **Succeeds** <br />
+> This example succeeds because `/users/item` is imported, making it available on the Graph Client
+
+```typescript
+import "@microsoft/microsoft-graph-client/users/item";
+const directReports = await graphClient.usersById("ab677c1b-50c6-4013-aa9f-792ed42c59c8").directReports.get();
+const extensions = await graphClient.usersById("ab677c1b-50c6-4013-aa9f-792ed42c59c8").extensions.get();
```

Comment:
/item/extensions hasn't been imported, this should fail too.

Reply:
Extensions is a child resource of /users/item, so I expect extensions to be available. Am I seeing this wrong? I'm not asking for a specific extension, but the collection.

Reply:
I had the same understanding as Sébastien here. Importing a node of the tree makes all of its direct children available.

Reply:
I'm trying to understand the expected behaviour. If we're [expecting the first level nodes to be automatically imported for discoverability](https://github.com/microsoftgraph/msgraph-sdk-javascript/pull/1104#discussion_r1050086783) and  child nodes for any nodes to be imported automatically as long as the parent is imported, we're effectively pulling the full tree and not doing any selective imports anymore.

Reply:
My understanding was that we'd be importing a node and that would make it's children, and no deeper in the tree available, e.g. importing `/users` makes `/users/{id}` available, but not `/users/{id}/directReports` as this would require `/users/item` to be imported, which would also make `/users/{id}/extensions` available.

Similarly, the root node makes all first level paths available, and in that case that node is already imported implicitly.